x-pack/plugins/security/public/hacks/on_session_timeout.js

```diff
@@ -26,50 +29,49 @@ module.config(($httpProvider) => {
     $q,
     $injector,
     sessionTimeout,
-    Notifier,
     Private,
     autoLogout
   ) => {
+
+    function refreshSession() {
+      // Make a simple request to keep the session alive
+      $injector.get('es').ping();
+      clearNotifications();
+    }
+
     const isUnauthenticated = Private(PathProvider).isUnauthenticated();
-    const notifier = new Notifier();
     const notificationLifetime = 60 * 1000;
     const notificationOptions = {
-      type: 'warning',
-      content: i18n.translate('xpack.security.hacks.logoutNotification', {
-        defaultMessage: 'You will soon be logged out due to inactivity. Click OK to resume.'
-      }),
-      icon: 'warning',
+      color: 'warning',
+      text: (
+        <SessionExpirationWarning onRefreshSession={refreshSession} />
+      ),
       title: i18n.translate('xpack.security.hacks.warningTitle', {
         defaultMessage: 'Warning'
       }),
-      lifetime: Math.min(
+      toastLifeTimeMs: Math.min(
         (sessionTimeout - SESSION_TIMEOUT_GRACE_PERIOD_MS),
         notificationLifetime
       ),
-      actions: ['accept']
     };
 
     let pendingNotification;
     let activeNotification;
+    let pendingSessionExpiration;
 
     function clearNotifications() {
       if (pendingNotification) $timeout.cancel(pendingNotification);
-      if (activeNotification) activeNotification.clear();
+      if (pendingSessionExpiration) clearTimeout(pendingSessionExpiration);
+      if (activeNotification) toastNotifications.remove(activeNotification);
     }
 
     function scheduleNotification() {
       pendingNotification = $timeout(showNotification, Math.max(sessionTimeout - notificationLifetime, 0));
     }
 
     function showNotification() {
-      activeNotification = notifier.add(notificationOptions, (action) => {
-        if (action === 'accept') {
-          // Make a simple request to keep the session alive
-          $injector.get('es').ping();
-        } else {
-          autoLogout();
-        }
-      });
+      activeNotification = toastNotifications.add(notificationOptions);
+      pendingSessionExpiration = setTimeout(() => autoLogout(), notificationOptions.toastLifeTimeMs);
```

Comment:
Did we used to be doing this elsewhere, or is this new?

Reply:
The old notification system invoked the action callback if the timeout expired. The `else` clause here:
```
if (action === 'accept') {
  // Make a simple request to keep the session alive
  $injector.get('es').ping();
} else {
  autoLogout();
}
```

Toast notifications don't have a callback to notify us when they reach their timeout, so this just implements that behavior.

Reply:
Gotcha, thanks for the explanation!